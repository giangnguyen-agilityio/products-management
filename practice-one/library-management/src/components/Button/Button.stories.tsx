import {type StoryFn, type Meta} from '@storybook/react'

import Button from './index'

export default {
  title: 'Example/Button',
  component: Button,
  argTypes: {
    size: {
      control: {
        type: 'select',
        options: ['small', 'medium', 'large'],
      },
    },
    variant: {
      control: {
        type: 'select',
        options: [
          'primary',
          'secondary',
          'default',
          'success',
          'warning',
          'danger',
        ],
      },
    },
    onClick: {action: 'clicked'},
  },
} as Meta<typeof Button>

const Template: StoryFn<typeof Button> = args => <Button {...args} />

export const Default = Template.bind({})
Default.args = {
  children: 'Button',
}

export const Disabled = Template.bind({})
Disabled.args = {
  children: 'Disabled Button',
  disabled: true,
}
