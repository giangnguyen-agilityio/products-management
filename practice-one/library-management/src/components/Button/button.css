.btn {
  display: flex;
  justify-content: center;
  align-items: center;
  border: none;
  font-family: var(--primary-font);
  color: var(--secondary-text-color);
  background-color: var(--primary-color);
  cursor: pointer;
  border-radius: 4px;
  padding: 10px 16px;
}

.btn:hover {
  opacity: 0.8;
}

.btn:disabled {
  cursor: default;
  color: var(--primary-text-color);
  background-color: var(--secondary-color);
}

.btn-default {
  background-color: var(--tertiary-color);
}

.btn-primary {
  background-color: var(--primary-color);
}

.btn-secondary {
  background-color: var(--secondary-color);
  color: var(--secondary-text-color);
  box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
}

.btn-success {
  background-color: var(--success-color);
}

.btn-warning {
  background-color: var(--warning-color);
}

.btn-danger {
  background-color: var(--danger-color);
}

.btn-small {
  font-size: 12px;
  padding: 10px 16px;
}

.btn-medium {
  font-size: 14px;
  padding: 11px 20px;
  font-weight: var(--secondary-font);
}

.btn-large {
  font-size: 16px;
  padding: 12px 24px;
  font-weight: var(--tertiary-font);
}
