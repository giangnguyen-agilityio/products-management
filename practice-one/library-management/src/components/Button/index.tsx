import './button.css'
interface ButtonProps {
  size?: 'small' | 'medium' | 'large'
  variant?:
  | 'primary'
  | 'secondary'
  | 'default'
  | 'success'
  | 'warning'
  | 'danger'
  className?: string
  ariaLabel?: string
  onClick?: () => void
}

const Button = (
  props: ButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>
): JSX.Element => {
  const { children, size, variant, className, ariaLabel, onClick, ...rest } =
    props

  return (
    <button
      className={`btn btn-${variant} btn-${size} ${className}`}
      onClick={onClick}
      type="button"
      aria-label={ariaLabel}
      {...rest}
    >
      {children}
    </button>
  )
}

export default Button
