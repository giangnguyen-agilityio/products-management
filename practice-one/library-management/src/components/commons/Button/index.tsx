import React, {memo} from 'react'
import './button.css'

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  size?: 'small' | 'medium' | 'large'
  variant?:
    | 'primary'
    | 'secondary'
    | 'default'
    | 'success'
    | 'warning'
    | 'danger'
  className: string
  ariaLabel?: string
  onClick?: () => void
  disabled?: boolean
  children?: React.ReactNode | undefined
}

const Button: React.FC<ButtonProps> = props => {
  // Destructure the props
  const {children, size, variant, className, ariaLabel, onClick, disabled} =
    props

  return (
    <button
      className={`btn btn-${variant || 'primary'} btn-${
        size || 'medium'
      } ${className}`}
      onClick={onClick}
      type="button"
      aria-label={ariaLabel}
      disabled={disabled}
    >
      {children}
    </button>
  )
}

export default memo(Button)
