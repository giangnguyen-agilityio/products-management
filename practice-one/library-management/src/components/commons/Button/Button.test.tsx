import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import {AiFillCheckCircle} from 'react-icons/ai'
import Button from '.'

describe('Button Component', () => {
  it('renders a button with correct default variant and size', () => {
    const {getByRole} = render(<Button />)
    const button = getByRole('button')

    expect(button).toMatchSnapshot()
  })

  it('renders with the correct icon', () => {
    const {queryByRole} = render(
      <Button>
        <AiFillCheckCircle />
      </Button>
    )
    const buttonElement = queryByRole('button')
    expect(buttonElement).toBeInTheDocument()
    if (buttonElement) {
      expect(buttonElement).toContainElement(document.querySelector('svg')!)
    }
  })

  it('renders a button with custom variant and size', () => {
    const {getByRole} = render(<Button variant="secondary" size="large" />)
    const button = getByRole('button')

    expect(button).toMatchSnapshot()
  })

  it('calls onClick when button is clicked', () => {
    const onClickMock = jest.fn()
    const {getByRole} = render(<Button onClick={onClickMock} />)
    const button = getByRole('button')

    fireEvent.click(button)

    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  it('renders a disabled button', () => {
    const {getByRole} = render(<Button disabled />)
    const button = getByRole('button')

    expect(button).toBeDisabled()
  })

  it('renders button with children', () => {
    const buttonText = 'Click Me'
    const {getByText} = render(<Button>{buttonText}</Button>)
    const button = getByText(buttonText)

    expect(button).toBeInTheDocument()
  })

  it('renders button with custom size class name', () => {
    const size = 'small'
    const {getByRole} = render(<Button size={size} />)
    const button = getByRole('button')

    expect(button).toHaveClass(`btn-${size}`)
  })

  it('renders button with custom variant class name', () => {
    const variant = 'warning'
    const {getByRole} = render(<Button variant={variant} />)
    const button = getByRole('button')

    expect(button).toHaveClass(`btn-${variant}`)
  })

  it('renders without error', () => {
    render(<Button />)
  })

  it('renders button with default props', () => {
    const {getByRole} = render(<Button />)
    const button = getByRole('button')

    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('btn')
    expect(button).toHaveClass('btn-primary')
    expect(button).toHaveClass('btn-medium')
    expect(button).not.toHaveAttribute('disabled')
  })

  it('renders button with custom size and variant', () => {
    const {getByRole} = render(
      <Button size="large" variant="secondary">
        Custom Button
      </Button>
    )
    const button = getByRole('button')

    expect(button).toBeInTheDocument()
    expect(button).toHaveClass('btn')
    expect(button).toHaveClass('btn-secondary')
    expect(button).toHaveClass('btn-large')
    expect(button).not.toHaveAttribute('disabled')
  })

  it('triggers onClick handler when clicked', () => {
    const handleClick = jest.fn()
    const {getByRole} = render(<Button onClick={handleClick}>Click me</Button>)
    const button = getByRole('button')

    fireEvent.click(button)

    expect(handleClick).toHaveBeenCalledTimes(1)
  })
})
