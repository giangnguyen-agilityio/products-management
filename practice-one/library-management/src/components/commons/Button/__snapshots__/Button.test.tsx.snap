// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders a button with correct default variant and size 1`] = `
<button
  class="btn btn-primary btn-medium "
  type="button"
/>
`;

exports[`Button Component renders a button with custom variant and size 1`] = `
<button
  class="btn btn-secondary btn-large "
  type="button"
/>
`;
